File: frontend/src/api/memory.ts
/**
 * Memory API — AI记忆系统API模块
 */
import api from './index'

export interface MemoryEntry {
  id: number
  workspace_id: number
  project_id?: number
  issue_id?: number
  agent_id?: number
  memory_type: 'short_term' | 'medium_term' | 'long_term'
  scope: 'workspace' | 'project' | 'issue' | 'agent'
  content: string
  embedding?: number[]
  metadata?: Record<string, any>
  tags?: string[]
  context_key: string
  context_name?: string
  relevance_score: number
  expires_at?: string
  created_at: string
  updated_at: string
}

export interface MemorySession {
  id: string
  workspace_id: number
  session_type: string
  context_id?: string
  started_at: string
  updated_at: string
  closed_at?: string
  memory_count: number
}

export interface MemoryListFilters {
  project_id?: number
  issue_id?: number
  agent_id?: number
  memory_type?: string
  scope?: string
  context_key?: string
  tag?: string
  limit?: number
  offset?: number
}

// ==================== Memory CRUD ====================

export async function listMemories(
  workspaceId: number,
  filters?: MemoryListFilters
): Promise<MemoryEntry[]> {
  const params: Record<string, any> = {}
  if (filters) {
    if (filters.project_id) params.project_id = filters.project_id
    if (filters.issue_id) params.issue_id = filters.issue_id
    if (filters.agent_id) params.agent_id = filters.agent_id
    if (filters.memory_type) params.memory_type = filters.memory_type
    if (filters.scope) params.scope = filters.scope
    if (filters.context_key) params.context_key = filters.context_key
    if (filters.tag) params.tag = filters.tag
    if (filters.limit) params.limit = filters.limit
    if (filters.offset) params.offset = filters.offset
  }
  const response = await api.get(`/workspaces/${workspaceId}/memories`, { params })
  return response.data
}

export async function getMemory(workspaceId: number, memoryId: number): Promise<MemoryEntry> {
  const response = await api.get(`/workspaces/${workspaceId}/memories/${memoryId}`)
  return response.data
}

export async function createMemory(
  workspaceId: number,
  data: Partial<MemoryEntry>
): Promise<MemoryEntry> {
  const response = await api.post(`/workspaces/${workspaceId}/memories`, data)
  return response.data
}

export async function updateMemory(
  workspaceId: number,
  memoryId: number,
  data: Partial<MemoryEntry>
): Promise<MemoryEntry> {
  const response = await api.put(`/workspaces/${workspaceId}/memories/${memoryId}`, data)
  return response.data
}

export async function deleteMemory(workspaceId: number, memoryId: number): Promise<void> {
  await api.delete(`/workspaces/${workspaceId}/memories/${memoryId}`)
}

// ==================== Memory Search ====================

export async function searchMemories(
  workspaceId: number,
  query: string,
  limit?: number
): Promise<MemoryEntry[]> {
  const response = await api.post(`/workspaces/${workspaceId}/memories/search`, {
    query,
    limit: limit || 10,
  })
  return response.data
}

export async function semanticSearchMemories(
  workspaceId: number,
  embedding: number[],
  limit?: number
): Promise<MemoryEntry[]> {
  const response = await api.post(`/workspaces/${workspaceId}/memories/semantic-search`, {
    embedding,
    limit: limit || 10,
  })
  return response.data
}

export async function getContextMemories(
  workspaceId: number,
  contextKey: string
): Promise<MemoryEntry[]> {
  const response = await api.get(`/workspaces/${workspaceId}/memories/context/${contextKey}`)
  return response.data
}

// ==================== Memory Sessions ====================

export async function createMemorySession(
  workspaceId: number,
  sessionType?: string
): Promise<MemorySession> {
  const response = await api.post(`/workspaces/${workspaceId}/memory-sessions`, {
    session_type: sessionType || 'chat',
  })
  return response.data
}

export async function getMemorySession(
  workspaceId: number,
  sessionId: string
): Promise<MemorySession> {
  const response = await api.get(`/workspaces/${workspaceId}/memory-sessions/${sessionId}`)
  return response.data
}

export async function closeMemorySession(workspaceId: number, sessionId: string): Promise<void> {
  await api.put(`/workspaces/${workspaceId}/memory-sessions/${sessionId}/close`)
}

export default {
  listMemories,
  getMemory,
  createMemory,
  updateMemory,
  deleteMemory,
  searchMemories,
  semanticSearchMemories,
  getContextMemories,
  createMemorySession,
  getMemorySession,
  closeMemorySession,
}

File: backend/internal/model/memory.go
package model

import (
	"encoding/json"
	"time"
)

// MemoryType defines the scope of memory
type MemoryType string

const (
	MemoryShortTerm  MemoryType = "short_term"   // Session-level, expires quickly
	MemoryMediumTerm MemoryType = "medium_term"  // Project-level, persists across sessions
	MemoryLongTerm   MemoryType = "long_term"    // Workspace-level, persists indefinitely
)

// MemoryScope defines where memory is applicable
type MemoryScope string

const (
	ScopeWorkspace MemoryScope = "workspace"
	ScopeProject   MemoryScope = "project"
	ScopeIssue     MemoryScope = "issue"
	ScopeAgent     MemoryScope = "agent"
)

// MemoryEntry represents a single memory entry
type MemoryEntry struct {
	BaseModel

	WorkspaceID uint64          `gorm:"not null;index" json:"workspace_id"`
	ProjectID   *uint64         `gorm:"index" json:"project_id,omitempty"`
	IssueID     *uint64         `gorm:"index" json:"issue_id,omitempty"`
	AgentID     *uint64         `gorm:"index" json:"agent_id,omitempty"`

	// Memory classification
	MemoryType MemoryType `gorm:"size:20;not null;index" json:"memory_type"`
	Scope      MemoryScope `gorm:"size:20;not null" json:"scope"`

	// Content
	Content     string          `gorm:"type:text;not null" json:"content"`
	Summary     string          `gorm:"-" json:"summary,omitempty"` // Generated summary for search results, not stored in DB
	Embedding   json.RawMessage `gorm:"type:jsonb" json:"embedding,omitempty"` // Vector embedding for semantic search
	Metadata    json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"metadata"`
	Tags        json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"tags"` // Array of strings

	// Context
	ContextKey  string `gorm:"size:255;index" json:"context_key"`   // Unique key for context grouping
	ContextName string `gorm:"size:255" json:"context_name"`        // Human-readable context name

	// Relevance scoring
	RelevanceScore float64 `gorm:"type:decimal(5,4);default:0" json:"relevance_score"`

	// Expiry
	ExpiresAt *time.Time `json:"expires_at,omitempty"` // For short-term memory

	// Relationships
	Workspace Workspace `gorm:"foreignKey:WorkspaceID" json:"-"`
	Project   *Project  `gorm:"foreignKey:ProjectID" json:"-"`
	Issue     *Issue    `gorm:"foreignKey:IssueID" json:"-"`
	Agent     *Agent    `gorm:"foreignKey:AgentID" json:"-"`
}

func (MemoryEntry) TableName() string { return "memory_entries" }

// MemorySession tracks the memory usage during a session
type MemorySession struct {
	ID          string    `gorm:"primaryKey;size:64" json:"id"`
	WorkspaceID uint64    `gorm:"not null;index" json:"workspace_id"`
	SessionType string    `gorm:"size:50;not null" json:"session_type"` // "chat", "task", "analysis", "agent"
	ContextID   *string   `gorm:"size:64" json:"context_id,omitempty"`
	StartedAt   time.Time `gorm:"autoCreateTime" json:"started_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	ClosedAt    *time.Time `json:"closed_at,omitempty"`
	MemoryCount int       `gorm:"default:0" json:"memory_count"`
}

func (MemorySession) TableName() string { return "memory_sessions" }

// MemoryVectorIndex provides inverted index for fast semantic search
type MemoryVectorIndex struct {
	ID           uint64    `gorm:"primaryKey" json:"id"`
	WorkspaceID  uint64    `gorm:"not null;index" json:"workspace_id"`
	MemoryEntryID uint64   `gorm:"not null;uniqueIndex" json:"memory_entry_id"`
	VectorData   json.RawMessage `gorm:"type:jsonb" json:"vector_data"` // Normalized vector
	CreatedAt    time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

func (MemoryVectorIndex) TableName() string { return "memory_vector_indexes" }